File: P17 Letter Combinations of a Phone Number/Solution.java
class Solution {
    // Time: O(2^n), where n is the length of digits
    // Space: O(n)
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if(digits.isEmpty()) {
            return result;
        }
        result.add("");
        String[] letterSets = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};     
        for(char c: digits.toCharArray()) {
            result = getCombinations(letterSets[Character.getNumericValue(c)], result);
        }
        return result;
    }
    
    private List<String> getCombinations(String possibleLetters, List<String> currentResult) {
            List<String> result = new ArrayList<String>();
            for (int i = 0; i < possibleLetters.length(); i++)  {
                for (String s : currentResult)  {
                    result.add(s + possibleLetters.charAt(i));
                }
            }
            return result;
        }
}

File: P17 Letter Combinations of a Phone Number/solution.py
class Solution:
    # Time: O(2^n), where n is the length of digits
    # Space: O(n)
    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        if not digits:
            return result
        result.append("")
        letter_sets = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        for i in range(len(digits)):
            result = self.get_combinations(letter_sets[ord(digits[i]) - ord('0')], result)
        return result
            
    def get_combinations(self, possible_letters: str, current_result: List[str])-> List[str]:
        result = []
        for i in range(len(possible_letters)):
            for s in current_result:
                result.append(s + possible_letters[i])
        return result
